include/benchmark/benchmark.h

```diff
@@ -180,17 +185,71 @@ void UseRealTime();
 namespace internal {
 class Benchmark;
 class BenchmarkFamilies;
+
+void StartBenchmarkTiming();
+void StopBenchmarkTiming();
 }
 
 // State is passed to a running Benchmark and contains state for the
 // benchmark to use.
 class State {
- public:
+public:
+  State(int max_iters, bool has_x, int x, bool has_y, int y, int thread_i)
+    : started_(false), total_iterations_(0), max_iterations_(max_iters),
+      has_range_x_(has_x), range_x_(x),
+      has_range_y_(has_y), range_y_(y),
+      bytes_processed_(0), items_processed_(0),
+      thread_index(thread_i)
+  {}
+
   // Returns true iff the benchmark should continue through another iteration.
-  bool KeepRunning();
+  BENCHMARK_ALWAYS_INLINE
+  bool KeepRunning() {
```

Comment:
I've done the legwork. Here are some of the things I noticed.

1. There was no difference when building in debug mode.
2. There was a difference when building in release. I saw a 5% change on some benchmarks in basic_test.

The one thing I like about moving `KeepRunning()` out of the header is that when the user compiles their tests with two different optimization levels the amount of overhead `KeepRunning()` introduces doesn't change.

However I also 

I think we can still expose fewer internals by moving some of the `State` methods out of the header. I'll make that change and not move `KeepRunning()` for now. I'll continue to look into this though.